Alerting: Add position-based matching for identical alert rules

konrad147 opened the pull request:
## Problem

When multiple identical alert rules exist in the same rule group (same name, labels, annotations, and query), the current matching algorithm cannot distinguish between them. This causes matching to fail and return `undefined`, even when the rules should match based on their position in the group.

This scenario occurs in real-world situations:
- Kubernetes monitoring with identical alerts for different namespaces
- Multi-tenant setups with replicated alert templates
- Rule synchronization between Prometheus and Ruler API

## Solution

Implemented position-based matching as a final fallback when all other matching criteria (name, labels, annotations, query) produce multiple identical matches.

### Key Changes

1. **Added `rulePositionHash` field** to `PromRuleWithOrigin` and `RulerRuleWithRulePosition`
   - Format: `"<index>:<totalRules>"` (e.g., `"0:3"`, `"1:5"`)
   - Encodes both the rule's position AND the total group size
   - Ensures groups must have the same structure to match

2. **Enhanced `getMatchingRulerRule()`** 
   - Now accepts `PromRuleWithOrigin` with position hash
   - Uses position as 4th-level tiebreaker for identical rules

3. **Enhanced `getMatchingPromRule()`**
   - Now accepts `RulerRuleWithRulePosition` with position hash
   - Provides bidirectional position-based matching

4. **Updated `matchRulesGroup()`**
   - Generates position hashes for ruler rules
   - Passes enhanced objects to matching functions

### Safety

Position matching only succeeds when:
- Rules are identical (name, labels, annotations, query)
- Both groups have the same number of rules (encoded in hash)
- Rules are at the same position

This prevents false positives when groups have different structures.

## Test Coverage

- Added 2 new edge case tests for position-based matching
- All 20 tests passing (up from 18 failing)
- Tests cover: unique names, labels/annotations, queries, position matching, and group size validation

## Files Changed

- `useFilteredRulesIterator.ts`: Added `rulePositionHash` to `PromRuleWithOrigin`
- `DataSourceRuleLoader.tsx`: Updated to pass full `PromRuleWithOrigin` object
- `FilterView.tsx`: Updated to pass full object to loader
- `ruleMatching.ts`: Enhanced both matching functions with position-based logic
- `ruleMatching.test.ts`: Added comprehensive test coverage

konrad147 committed:
Add rule index as a part of the rule matching process

gillesdemey commented on public/app/features/alerting/unified/rule-list/DataSourceGroupLoader.tsx:
```diff
@@ -194,43 +194,53 @@ export function RulerBasedGroupRules({
-            groupIdentifier={groupIdentifier}
+        if (promRule) {
+          // If rules are indistinguishable by name, labels, annotations, and query, we need to use the index to disambiguate
+          const promRuleIndex = promGroup.rules.indexOf(promRule);
```
What would happen if we just used the `index` from `pageItems`? Wouldn't that guarantee a unique key for each list item too?

gillesdemey reviewed (approved):
couple of small nits but overall great work! 🚀

soniaAguilarPeiron reviewed (approved):
LGTM!

konrad147 committed:
Refactor rule matching to use a dedicated RulePositionHash type

- Introduced a new `rulePositionHash.ts` file to define a branded type for rule positions.
- Updated rule matching logic to utilize the new `createRulePositionHash` function for generating rule position hashes.
- Adjusted related components and tests to ensure consistent handling of rule indices and positions.

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/esc-18904
